SU2_CFD/include/solvers/CFVMFlowSolverBase.hpp

```diff
@@ -949,36 +967,90 @@ class CFVMFlowSolverBase : public CSolver {
       Omega_Max = 0.0;
     }
     SU2_OMP_BARRIER
 
-    nodes->SetVorticity_StrainMag();
+    su2double strainMax = 0.0, omegaMax = 0.0;
 
-    /*--- Min and Max are not really differentiable ---*/
-    const bool wasActive = AD::BeginPassive();
+    SU2_OMP_FOR_STAT(omp_chunk_size)
+    for (unsigned long iPoint = 0; iPoint < nPoint; ++iPoint) {
 
-    su2double strainMax = 0.0, omegaMax = 0.0;
+      constexpr size_t u = VelocityOffset;
+      constexpr size_t v = VelocityOffset+1;
+      constexpr size_t w = VelocityOffset+2;
 
-    SU2_OMP(for schedule(static,omp_chunk_size) nowait)
-    for (unsigned long iPoint = 0; iPoint < nPoint; iPoint++) {
-      strainMax = max(strainMax, nodes->GetStrainMag(iPoint));
-      omegaMax = max(omegaMax, GeometryToolbox::Norm(3, nodes->GetVorticity(iPoint)));
-    }
-    SU2_OMP_CRITICAL {
-      StrainMag_Max = max(StrainMag_Max, strainMax);
-      Omega_Max = max(Omega_Max, omegaMax);
+      /*--- Vorticity ---*/
+
+      su2double* Vorticity = nodes->GetVorticity(iPoint);
+
+      Vorticity[0] = 0.0; Vorticity[1] = 0.0;
+
+      Vorticity[2] = Gradient_Primitive(iPoint,v,0)-Gradient_Primitive(iPoint,u,1);
+
+      if (nDim == 3) {
+        Vorticity[0] = Gradient_Primitive(iPoint,w,1)-Gradient_Primitive(iPoint,v,2);
+        Vorticity[1] = -(Gradient_Primitive(iPoint,w,0)-Gradient_Primitive(iPoint,u,2));
+      }
+
+      /*--- Strain Magnitude ---*/
+
+      AD::StartPreacc();
+      AD::SetPreaccIn(&Gradient_Primitive[iPoint][VelocityOffset], nDim, nDim);
+
+      su2double Div = 0.0;
+      for (unsigned long iDim = 0; iDim < nDim; iDim++)
+        Div += Gradient_Primitive(iPoint, iDim+VelocityOffset, iDim);
+      Div /= 3.0;
+
+      StrainMag(iPoint) = 0.0;
+
+      /*--- Add diagonal part ---*/
+
+      for (unsigned long iDim = 0; iDim < nDim; iDim++) {
+        StrainMag(iPoint) += pow(Gradient_Primitive(iPoint, iDim+VelocityOffset, iDim) - Div, 2);
+      }
+      if (nDim == 2) {
+        StrainMag(iPoint) += pow(Div, 2);
+      }
+
+      /*--- Add off diagonals ---*/
+
+      StrainMag(iPoint) += 2.0*pow(0.5*(Gradient_Primitive(iPoint,u,1) + Gradient_Primitive(iPoint,v,0)), 2);
+
+      if (nDim == 3) {
+        StrainMag(iPoint) += 2.0*pow(0.5*(Gradient_Primitive(iPoint,u,2) + Gradient_Primitive(iPoint,w,0)), 2);
+        StrainMag(iPoint) += 2.0*pow(0.5*(Gradient_Primitive(iPoint,v,2) + Gradient_Primitive(iPoint,w,1)), 2);
+      }
+
+      StrainMag(iPoint) = sqrt(2.0*StrainMag(iPoint));
+      AD::SetPreaccOut(StrainMag(iPoint));
+
+      /*--- Max is not differentiable, we so not register for preacc. ---*/
+      strainMax = max(strainMax, StrainMag(iPoint));
+      omegaMax = max(omegaMax, GeometryToolbox::Norm(3, Vorticity));
+
+      AD::EndPreacc();
     }
 
     if ((iMesh == MESH_0) && (config.GetComm_Level() == COMM_FULL)) {
+      SU2_OMP_CRITICAL {
+        StrainMag_Max = max(StrainMag_Max, strainMax);
+        Omega_Max = max(Omega_Max, omegaMax);
+      }
+
```

Comment:
I suppose you do the reduction over the threads yourself because of efficiency, when the if statement is false, but wouldn't it be more readable to have a reduction clause in the OMP for loop?

Reply:
Thank you for having a look at the code Edwin, the issues with these kind of reductions were:
- The microsoft compilers do not support max/min reductions;
- For the code to compile with AD (forward) we would need to define custom reduction operators, which might not be very readable;
- Earlier discussions about OpenMP+AD (reverse) suggested that avoiding reduction clauses would make it easier to develop a thread-safe AD tool.

For performance the reduction clause would be more efficient, as the compiler can make min/max atomic, however I do not know of a nice simple way to implement atomic min/max.